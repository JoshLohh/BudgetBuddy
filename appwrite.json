{
    "projectId": "683b7b97000aae994ff5",
    "projectName": "BudgetBuddy",
    "settings": {
        "services": {
            "account": true,
            "avatars": true,
            "databases": true,
            "locale": true,
            "health": true,
            "storage": true,
            "teams": true,
            "users": true,
            "sites": true,
            "functions": true,
            "graphql": true,
            "messaging": true
        },
        "auth": {
            "methods": {
                "jwt": true,
                "phone": true,
                "invites": true,
                "anonymous": true,
                "email-otp": true,
                "magic-url": true,
                "email-password": true
            },
            "security": {
                "duration": 31536000,
                "limit": 0,
                "sessionsLimit": 10,
                "passwordHistory": 0,
                "passwordDictionary": false,
                "personalDataCheck": false,
                "sessionAlerts": false,
                "mockNumbers": []
            }
        }
    },
    "functions": [
        {
            "$id": "685d1db2003731b3c3d6",
            "execute": [
                "any"
            ],
            "name": "CreateUserProfile",
            "enabled": true,
            "logging": true,
            "runtime": "node-18.0",
            "scopes": [
                "users.read",
                "databases.read",
                "databases.write",
                "collections.read",
                "collections.write",
                "attributes.read",
                "attributes.write",
                "indexes.read",
                "indexes.write",
                "documents.read",
                "documents.write"
            ],
            "events": [
                "users.create"
            ],
            "schedule": "",
            "timeout": 15,
            "entrypoint": "src/main.js",
            "commands": "npm install",
            "specification": "s-0.5vcpu-512mb",
            "path": "functions/CreateUserProfile"
        },
        {
            "$id": "685d26ee003371a58959",
            "execute": [
                "any"
            ],
            "name": "create-user",
            "enabled": true,
            "logging": true,
            "runtime": "bun-1.1",
            "scopes": [
                "users.read",
                "databases.read",
                "databases.write",
                "collections.read",
                "collections.write",
                "attributes.read",
                "attributes.write",
                "indexes.read",
                "indexes.write",
                "documents.read",
                "documents.write"
            ],
            "events": [
                "users.*.create"
            ],
            "schedule": "",
            "timeout": 15,
            "entrypoint": "src/main.ts",
            "commands": "bun install",
            "specification": "s-0.5vcpu-512mb",
            "path": "functions/create-user"
        }
    ],
    "databases": [
        {
            "$id": "684bd404003542c8a2ac",
            "name": "budget_buddy_db",
            "enabled": true
        }
    ],
    "collections": [
        {
            "$id": "684bd42e0009b1e585e7",
            "$permissions": [
                "create(\"users\")",
                "read(\"users\")",
                "update(\"users\")",
                "delete(\"users\")"
            ],
            "databaseId": "684bd404003542c8a2ac",
            "name": "groups",
            "enabled": true,
            "documentSecurity": true,
            "attributes": [
                {
                    "key": "title",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "members",
                    "type": "string",
                    "required": false,
                    "array": true,
                    "size": 500,
                    "default": null
                },
                {
                    "key": "createdBy",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "description",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 500,
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "684bd5ef0030ac0cc7f7",
            "$permissions": [],
            "databaseId": "684bd404003542c8a2ac",
            "name": "expenses",
            "enabled": true,
            "documentSecurity": false,
            "attributes": [
                {
                    "key": "groupId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "title",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "amount",
                    "type": "double",
                    "required": true,
                    "array": false,
                    "min": 0,
                    "max": 1.7976931348623157e+308,
                    "default": null
                },
                {
                    "key": "paidBy",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "splitBetween",
                    "type": "string",
                    "required": false,
                    "array": true,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "splitType",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 255,
                    "default": "equal"
                },
                {
                    "key": "customSplit",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 1000,
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "684bd8a5003d02f9b84d",
            "$permissions": [],
            "databaseId": "684bd404003542c8a2ac",
            "name": "settlements",
            "enabled": true,
            "documentSecurity": false,
            "attributes": [
                {
                    "key": "groupId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "from",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "to",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "amount",
                    "type": "double",
                    "required": true,
                    "array": false,
                    "min": 0,
                    "max": 1.7976931348623157e+308,
                    "default": null
                },
                {
                    "key": "date",
                    "type": "datetime",
                    "required": true,
                    "array": false,
                    "format": "",
                    "default": null
                }
            ],
            "indexes": []
        },
        {
            "$id": "685cf12e003976d23b39",
            "$permissions": [
                "create(\"any\")",
                "read(\"any\")",
                "update(\"any\")",
                "delete(\"any\")"
            ],
            "databaseId": "684bd404003542c8a2ac",
            "name": "users",
            "enabled": true,
            "documentSecurity": true,
            "attributes": [
                {
                    "key": "userId",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "username",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "email",
                    "type": "string",
                    "required": true,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "avatar",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 255,
                    "default": null
                },
                {
                    "key": "bio",
                    "type": "string",
                    "required": false,
                    "array": false,
                    "size": 1000,
                    "default": null
                }
            ],
            "indexes": []
        }
    ]
}