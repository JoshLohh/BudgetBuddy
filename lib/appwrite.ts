import { Client, Account, ID, Avatars } from 'react-native-appwrite';


export const client = new Client()
    .setEndpoint('https://fra.cloud.appwrite.io/v1')
    .setProject("683b7b97000aae994ff5")
    .setPlatform('dev.buddies.BudgetBuddy');

export const account = new Account(client)
export const avatars = new Avatars(client)