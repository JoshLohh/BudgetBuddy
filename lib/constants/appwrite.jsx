import { Client, Account, ID, Avatars } from 'react-native-appwrite';


export const client = new Client
    .setProject("6838ad0d003d2a0754a0")
    .setPlatform('dev.Buddies.budgetbuddy');

export const account = new Account(client)
export const avatars = new Avatars(client)