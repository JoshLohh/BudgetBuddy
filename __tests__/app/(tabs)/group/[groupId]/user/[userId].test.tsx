/**
 * Tests for UserProfileScreen component
 * - Mocks Appwrite SDK calls
 * - Mocks Expo Router hooks
 * - Checks data fetching, rendering, loading/error UI, and navigation
 */

import React from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react-native';
import UserProfileScreen from '@/app/(tabs)/group/[groupId]/user/[userId]';
import { useRouter, useLocalSearchParams } from 'expo-router';
import { databases } from '@/lib/appwrite';

// Mock Expo Router hooks
jest.mock('expo-router', () => ({
  useRouter: jest.fn(),
  useLocalSearchParams: jest.fn(),
}));

// Mock Appwrite databases client
jest.mock('@/lib/appwrite', () => ({
  databases: {
    getDocument: jest.fn(),
    listDocuments: jest.fn(),
  },
}));

describe('UserProfileScreen', () => {
  const mockBack = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    // Provide default mocks
    (useRouter as jest.Mock).mockReturnValue({ back: mockBack });
    (useLocalSearchParams as jest.Mock).mockReturnValue({ userId: 'test-user', groupId: 'test-group' });
  });

  it('renders loading indicator initially', () => {
    (databases.getDocument as jest.Mock).mockImplementation(() => new Promise(() => {})); // never resolves

    const { getByTestId } = render(<UserProfileScreen />);
    expect(getByTestId('loading-indicator')).toBeTruthy();
  });

  it('renders user profile and stats after successful fetch', async () => {
    (databases.getDocument as jest.Mock).mockResolvedValue({
      $id: 'test-user',
      username: 'Test User',
      avatar: 'avatar.png',
      bio: 'Hello bio',
      email: 'test@example.com',
    });
    (databases.listDocuments as jest.Mock)
      // groups query result
      .mockResolvedValueOnce({
        documents: [{ $id: 'group1' }, { $id: 'group2' }],
      })
      // expenses query result
      .mockResolvedValueOnce({
        documents: [
          { amount: '10.50' },
          { amount: '20' },
        ],
      });

    const { getAllByText, getByText, queryByTestId } = render(<UserProfileScreen />);
    
    // Wait for async updates
    await waitFor(() => {
      expect(queryByTestId('loading-indicator')).toBeNull(); // loading ended
    });

    // Check profile info rendered
    expect(getByText('Test User')).toBeTruthy();
    expect(getByText('Hello bio')).toBeTruthy();

    // Check stats rendered
    const allTwos = getAllByText('2');
    expect(allTwos.length).toBeGreaterThan(0);
    expect(getByText('$30.50')).toBeTruthy(); // total spent with decimals
  });

  it('renders "User not found" message on error', async () => {
    (databases.getDocument as jest.Mock).mockRejectedValue(new Error('Not found'));
    (databases.listDocuments as jest.Mock).mockResolvedValue({ documents: [] });

    const { getByText, queryByTestId } = render(<UserProfileScreen />);

    await waitFor(() => {
      expect(queryByTestId('loading-indicator')).toBeNull();
    });

    expect(getByText('User not found.')).toBeTruthy();
  });

  it('calls router.back() when back button pressed', async () => {
    (databases.getDocument as jest.Mock).mockResolvedValue({
      $id: 'test-user',
      username: 'Test User',
    });
    (databases.listDocuments as jest.Mock).mockResolvedValue({ documents: [] });

    const { getByTestId } = render(<UserProfileScreen />);
    await waitFor(() => {
      expect(getByTestId('back-button')).toBeTruthy();
    });

    fireEvent.press(getByTestId('back-button'));
    expect(mockBack).toHaveBeenCalled();
  });

  it('matches snapshot during loading', () => {
    (databases.getDocument as jest.Mock).mockImplementation(() => new Promise(() => {}));
    const { toJSON } = render(<UserProfileScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('matches snapshot with empty stats', async () => {
    (databases.getDocument as jest.Mock).mockResolvedValue({
      $id: 'test-user',
      username: 'Test User',
      avatar: 'avatar.png',
      bio: '',
      email: 'test@example.com',
    });
    (databases.listDocuments as jest.Mock)
      .mockResolvedValueOnce({ documents: [] }) // no groups
      .mockResolvedValueOnce({ documents: [] }); // no expenses

    const { queryByTestId, toJSON } = render(<UserProfileScreen />);

    await waitFor(() => {
      expect(queryByTestId('loading-indicator')).toBeNull();
    });

    expect(toJSON()).toMatchSnapshot();
  });
  
});
