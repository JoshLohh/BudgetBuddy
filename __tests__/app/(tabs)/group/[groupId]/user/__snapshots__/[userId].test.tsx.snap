// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileScreen matches snapshot during loading 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    testID="loading-indicator"
  />
</View>
`;

exports[`UserProfileScreen matches snapshot with empty stats 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "flex": 1,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingHorizontal": 20,
        "paddingTop": 36,
      },
    ]
  }
  testID="themed-view"
>
  <View
    style={
      {
        "height": 40,
        "width": "100%",
      }
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "left": 10,
        "opacity": 1,
        "padding": 8,
        "position": "absolute",
        "top": 60,
        "zIndex": 10,
      }
    }
    testID="back-button"
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": "#1976d2",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      î¨¨
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 18,
        "marginTop": 24,
      }
    }
  >
    <ViewManagerAdapter_ExpoImage
      borderColor={4294967295}
      borderRadius={50}
      borderWidth={2}
      containerViewRef={"[React.ref]"}
      contentFit="cover"
      contentPosition={
        {
          "left": "50%",
          "top": "50%",
        }
      }
      height={100}
      marginRight={18}
      nativeViewRef={"[React.ref]"}
      onError={[Function]}
      onLoad={[Function]}
      onLoadStart={[Function]}
      onProgress={[Function]}
      placeholder={[]}
      source={
        [
          {
            "uri": "avatar.png",
          },
        ]
      }
      style={
        {
          "borderColor": "#fff",
          "borderRadius": 50,
          "borderWidth": 2,
          "height": 100,
          "marginRight": 18,
          "width": 100,
        }
      }
      transition={null}
      width={100}
    />
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            undefined,
            {
              "fontSize": 28,
              "fontWeight": "bold",
              "lineHeight": 32,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 24,
              "fontWeight": "bold",
              "marginBottom": 0,
            },
          ]
        }
      >
        Test User
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "borderRadius": 12,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 24,
        "padding": 18,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        0
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Groups
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        $
        0.00
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Total Spent
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        0
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Expenses
      </Text>
    </View>
  </View>
</View>
`;
